server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -3083,21 +3084,44 @@ assert getActiveOperationsCount() == OPERATIONS_BLOCKED
         final TranslogStats translogStats = translogStats();
         // flush to make sure the latest commit, which will be opened by the read-only engine, includes all operations.
         flush(new FlushRequest().waitIfOngoing(true));
-        synchronized (mutex) {
-            verifyNotClosed();
-            // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
-            final Engine readOnlyEngine = new ReadOnlyEngine(newEngineConfig(), seqNoStats, translogStats, false, Function.identity());
-            IOUtils.close(currentEngineReference.getAndSet(readOnlyEngine));
-        }
 
         Engine newEngine = null;
+        SetOnce<Engine> newEngineReference = new SetOnce<>();
         try {
             final long globalCheckpoint = getGlobalCheckpoint();
             synchronized (mutex) {
-                assert currentEngineReference.get() instanceof ReadOnlyEngine : "another write engine is running";
                 verifyNotClosed();
                 // we must create a new engine under mutex (see IndexShard#snapshotStoreMetadata).
+                final Engine readOnlyEngine = new ReadOnlyEngine(newEngineConfig(), seqNoStats, translogStats, false, Function.identity()) {
+                    @Override
+                    public IndexCommitRef acquireLastIndexCommit(boolean flushFirst) {
+                        synchronized (mutex) {
+                            return newEngineReference.get().acquireLastIndexCommit(flushFirst);
```

Comment:
Here we should always pass `false` for `flushFirst` otherwise this method might throw Exception if the new engine hasn't finished its recovery yet.

Reply:
Fixed in 3409ed4ff4a867c9a978b1a5901d62c6fe318602